src/com/amazon/ion/impl/_Private_LSTWriter.java

```diff
@@ -0,0 +1,300 @@
+    private int depth;
+    private SymbolTable symbolTable;
+    private State state;
+    private LinkedList<SSTImportDescriptor> declaredImports;
+    private LinkedList<String> declaredSymbols;
+    private final IonCatalog catalog;
+    private boolean seenImports;
+    private boolean seenSymbols;
+
+    /**
+     * Enum for managing our state between API calls.
+     * <p>
+     * LST_STRUCT: we are in an lst struct at depth 1.
+     * PRE_IMPORT_LIST: setfield was called with the symbol 'imports' from LST_STRUCT.
+     * IMPORT_LIST: either stepin(List) was called from PRE_IMPORT_LIST, or stepOut from IMPORT_DESCRIPTOR.
+     * IMPORT_DESCRIPTOR: StepIn(Struct) from IMPORT_LIST was called or a field/value pair was written.
+     * IMPORT_MAX_ID: setfield was called with the symbol 'max_id' from IMPORT_DESCRIPTOR.
+     * IMPORT_NAME: setfield was called with the symbol 'name' from IMPORT_DESCRIPTOR.
+     * IMPORT_VERSION: setfield was called with the symbol 'version' from IMPORT_DESCRIPTOR.
+     * PRE_SYMBOL: setfield was called with the symbol 'symbols' from LST_STRUCT.
+     * SYMBOL_LIST: stepin(List) was called from PRE_SYMBOL_LIST.
+     */
+    private enum State {LST_STRUCT, PRE_IMPORT_LIST, IMPORT_LIST, IMPORT_DESCRIPTOR, IMPORT_MAX_ID, IMPORT_NAME, IMPORT_VERSION, PRE_SYMBOL_LIST, SYMBOL_LIST}
+
+    /**
+     * Holds import descriptor info between API calls.
+     **/
+    private static class SSTImportDescriptor {
+        String name;
+        int version;
+        int maxID;
+    }
+
+    /**
+     * Constructor for the _Private_LSTWriter
+     * <p>
+     * Absorbs system processing from user level APIs in the IonManagedBinaryWriter.
+     *
+     * @param  imports ArrayList<SymbolTable> is populated by the builder.
+     * @param  symbols List<String>
+     * @param  inCatalog IonCatalog
+     */
+   public _Private_LSTWriter(List<SymbolTable> imports, List<String> symbols, IonCatalog inCatalog) {
+        catalog = inCatalog;
+        declaredSymbols = new LinkedList<String>();
+        if(imports.isEmpty() || !imports.get(0).isSystemTable()){
+            imports.add(0, _Private_Utils.systemSymtab(1));
+        }
+        symbolTable = new LocalSymbolTable(new LocalSymbolTableImports(imports), symbols);
+        depth = 0;
+    }
+
+    public SymbolTable getSymbolTable() {
+        return symbolTable;
+    }
+
+    public void stepIn(IonType containerType) throws IOException {
+        if(state == null) {
+            if(depth == 0) {
+                state = State.LST_STRUCT;
+            } else {
+                throw new UnsupportedOperationException("$ion_symbol_table:: was not added to a struct.");
+            }
+        }
+        depth++;
+        switch(state) {
+            case LST_STRUCT:
+                return;
+            case PRE_IMPORT_LIST:
+                //entering the declared Imports list.
+                state = State.IMPORT_LIST;
```

Comment:
shouldn't you check that the container type is list as well? I think this would pass for `import: ()`. Same for the other states